File: data_sources/ais/db_synthetic/check_intersections.py
#!/usr/bin/env python3
"""
Script to check if trajectory crosses any infrastructure cables

Trajectory coordinates (lat, lon):
- 54.3374597, 12.0771658
- 54.3310199, 11.9667175
- 54.3353132, 11.8415427
"""

import json
import math
from typing import List, Tuple, Dict, Optional

Point = Tuple[float, float]  # (lon, lat)
LineSegment = Tuple[Point, Point]


def cross_product(o: Point, a: Point, b: Point) -> float:
    """Calculate cross product of vectors oa and ob"""
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])


def on_segment(p: Point, q: Point, r: Point) -> bool:
    """Check if point q lies on segment pr"""
    return (
        q[0] <= max(p[0], r[0]) and
        q[0] >= min(p[0], r[0]) and
        q[1] <= max(p[1], r[1]) and
        q[1] >= min(p[1], r[1])
    )


def segments_intersect(seg1: LineSegment, seg2: LineSegment) -> bool:
    """Check if two line segments intersect using cross product method"""
    a, b = seg1
    c, d = seg2

    # Calculate cross products
    o1 = cross_product(a, c, d)
    o2 = cross_product(b, c, d)
    o3 = cross_product(c, a, b)
    o4 = cross_product(d, a, b)

    # General case: segments intersect if points are on opposite sides
    if o1 * o2 < 0 and o3 * o4 < 0:
        return True

    # Special cases: check if endpoints are collinear and on the segment
    if o1 == 0 and on_segment(c, a, d):
        return True
    if o2 == 0 and on_segment(c, b, d):
        return True
    if o3 == 0 and on_segment(a, c, b):
        return True
    if o4 == 0 and on_segment(a, d, b):
        return True

    return False


def distance(p1: Point, p2: Point) -> float:
    """Calculate distance between two points"""
    dlon = p2[0] - p1[0]
    dlat = p2[1] - p1[1]
    return math.sqrt(dlon * dlon + dlat * dlat)


def point_to_segment_distance(point: Point, segment: LineSegment) -> float:
    """Calculate minimum distance from point to line segment"""
    p1, p2 = segment
    
    A = point[0] - p1[0]
    B = point[1] - p1[1]
    C = p2[0] - p1[0]
    D = p2[1] - p1[1]

    dot = A * C + B * D
    len_sq = C * C + D * D
    
    if len_sq == 0:
        return distance(point, p1)
    
    param = dot / len_sq

    if param < 0:
        xx, yy = p1
    elif param > 1:
        xx, yy = p2
    else:
        xx = p1[0] + param * C
        yy = p1[1] + param * D

    dx = point[0] - xx
    dy = point[1] - yy
    return math.sqrt(dx * dx + dy * dy)


def check_intersections():
    """Main function to check for intersections"""
    print("🔍 Checking for infrastructure intersections with trajectory...\n")

    # Load infrastructure data
    with open('../../infrastructure/internet_cables_coordinates.json', 'r') as f:
        infrastructure_data = json.load(f)

    # Define trajectory segments (converted from lat,lon to lon,lat)
    trajectory_points: List[Point] = [
        (12.0771658, 54.3374597),
        (11.9667175, 54.3310199),
        (11.8415427, 54.3353132),
    ]

    trajectory_segments: List[LineSegment] = [
        (trajectory_points[0], trajectory_points[1]),
        (trajectory_points[1], trajectory_points[2]),
    ]

    print("Trajectory segments:")
    for i, seg in enumerate(trajectory_segments):
        print(f"  Segment {i + 1}: ({seg[0][0]}, {seg[0][1]}) → ({seg[1][0]}, {seg[1][1]})")
    print()

    intersections: List[Dict] = []
    nearby_cables: List[Dict] = []

    # Check each cable
    for feature in infrastructure_data['features']:
        cable_name = feature['properties']['name']
        cable_id = feature['properties']['id']
        geometry = feature['geometry']

        if geometry['type'] != 'MultiLineString':
            continue

        # Check each line string in the MultiLineString
        for seg_idx, line_string in enumerate(geometry['coordinates']):
            # Check each segment in the line string
            for i in range(len(line_string) - 1):
                cable_segment: LineSegment = (
                    (line_string[i][0], line_string[i][1]),
                    (line_string[i + 1][0], line_string[i + 1][1])
                )

                # Check intersection with each trajectory segment
                for traj_idx, traj_segment in enumerate(trajectory_segments):
                    if segments_intersect(traj_segment, cable_segment):
                        intersections.append({
                            'cableName': cable_name,
                            'cableId': cable_id,
                            'segmentIndex': seg_idx,
                            'trajectorySegment': traj_idx + 1,
                        })

                # Also check if trajectory points are near this cable segment
                for traj_point in trajectory_points:
                    dist = point_to_segment_distance(traj_point, cable_segment)
                    # 0.05 degrees ≈ 5.5 km
                    if dist < 0.05:
                        nearby_cables.append({
                            'cableName': cable_name,
                            'cableId': cable_id,
                            'minDistance': dist,
                            'closestPoint': traj_point,
                        })

    # Report results
    print("📊 Results:\n")

    if intersections:
        print(f"✅ Found {len(intersections)} intersection(s):\n")
        unique_cables = {}
        for intersection in intersections:
            key = f"{intersection['cableId']}-{intersection['segmentIndex']}"
            if key not in unique_cables:
                unique_cables[key] = intersection

        for intersection in unique_cables.values():
            print(f"  🔴 {intersection['cableName']} ({intersection['cableId']})")
            print(f"     Intersects with trajectory segment {intersection['trajectorySegment']}")
            print(f"     Cable segment index: {intersection['segmentIndex']}\n")
    else:
        print("❌ No direct intersections found.\n")

    # Check for nearby cables (within 0.05 degrees ≈ 5.5 km)
    unique_nearby = {}
    if nearby_cables:
        print(f"⚠️  Found {len(nearby_cables)} nearby cable detection(s) (within ~5.5 km):\n")
        for cable in nearby_cables:
            key = cable['cableId']
            if key not in unique_nearby or unique_nearby[key]['minDistance'] > cable['minDistance']:
                unique_nearby[key] = cable

        for cable in unique_nearby.values():
            # Convert degrees to km (roughly 111 km per degree)
            dist_km = cable['minDistance'] * 111
            print(f"  ⚠️  {cable['cableName']} ({cable['cableId']})")
            print(f"     Minimum distance: ~{dist_km:.2f} km\n")

    return {
        'intersections': list(unique_cables.values()) if intersections else [],
        'nearbyCables': list(unique_nearby.values()),
    }


if __name__ == '__main__':
    try:
        result = check_intersections()
        print("\n✅ Analysis complete.")
    except Exception as error:
        print(f"❌ Error: {error}")
        import traceback
        traceback.print_exc()
        exit(1)

File: data_sources/ais/db_synthetic/check_intersections.ts
/**
 * Script to check if trajectory crosses any infrastructure cables
 * 
 * Trajectory coordinates (lat, lon):
 * - 54.3374597, 12.0771658
 * - 54.3310199, 11.9667175
 * - 54.3353132, 11.8415427
 */

interface Point {
  lon: number;
  lat: number;
}

interface LineSegment {
  p1: Point;
  p2: Point;
}

/**
 * Check if two line segments intersect
 * Uses the cross product method for line segment intersection
 */
function segmentsIntersect(seg1: LineSegment, seg2: LineSegment): boolean {
  const { p1: a, p2: b } = seg1;
  const { p1: c, p2: d } = seg2;

  // Helper function to calculate cross product
  const crossProduct = (o: Point, a: Point, b: Point): number => {
    return (a.lon - o.lon) * (b.lat - o.lat) - (a.lat - o.lat) * (b.lon - o.lon);
  };

  // Check if points are on opposite sides of each segment
  const o1 = crossProduct(a, c, d);
  const o2 = crossProduct(b, c, d);
  const o3 = crossProduct(c, a, b);
  const o4 = crossProduct(d, a, b);

  // General case: segments intersect if points are on opposite sides
  if (o1 * o2 < 0 && o3 * o4 < 0) {
    return true;
  }

  // Special cases: check if endpoints are collinear and on the segment
  if (o1 === 0 && onSegment(c, a, d)) return true;
  if (o2 === 0 && onSegment(c, b, d)) return true;
  if (o3 === 0 && onSegment(a, c, b)) return true;
  if (o4 === 0 && onSegment(a, d, b)) return true;

  return false;
}

/**
 * Check if point q lies on segment pr
 */
function onSegment(p: Point, q: Point, r: Point): boolean {
  return (
    q.lon <= Math.max(p.lon, r.lon) &&
    q.lon >= Math.min(p.lon, r.lon) &&
    q.lat <= Math.max(p.lat, r.lat) &&
    q.lat >= Math.min(p.lat, r.lat)
  );
}

/**
 * Calculate distance between two points (for proximity check)
 */
function distance(p1: Point, p2: Point): number {
  const dlon = p2.lon - p1.lon;
  const dlat = p2.lat - p1.lat;
  return Math.sqrt(dlon * dlon + dlat * dlat);
}

/**
 * Check if a point is within a certain distance of a line segment
 */
function pointNearSegment(point: Point, segment: LineSegment, threshold: number = 0.01): boolean {
  const { p1, p2 } = segment;
  
  // Calculate distance from point to line segment
  const A = point.lon - p1.lon;
  const B = point.lat - p1.lat;
  const C = p2.lon - p1.lon;
  const D = p2.lat - p1.lat;

  const dot = A * C + B * D;
  const lenSq = C * C + D * D;
  let param = -1;
  
  if (lenSq !== 0) {
    param = dot / lenSq;
  }

  let xx: number, yy: number;

  if (param < 0) {
    xx = p1.lon;
    yy = p1.lat;
  } else if (param > 1) {
    xx = p2.lon;
    yy = p2.lat;
  } else {
    xx = p1.lon + param * C;
    yy = p1.lat + param * D;
  }

  const dx = point.lon - xx;
  const dy = point.lat - yy;
  const dist = Math.sqrt(dx * dx + dy * dy);
  
  return dist <= threshold;
}

async function checkIntersections() {
  console.log('🔍 Checking for infrastructure intersections with trajectory...\n');

  // Load infrastructure data
  const infrastructureFile = Bun.file(
    '../infrastructure/internet_cables_coordinates.json'
  );
  const infrastructureData = await infrastructureFile.json();

  // Define trajectory segments (converted from lat,lon to lon,lat)
  const trajectoryPoints: Point[] = [
    { lon: 12.0771658, lat: 54.3374597 },
    { lon: 11.9667175, lat: 54.3310199 },
    { lon: 11.8415427, lat: 54.3353132 },
  ];

  const trajectorySegments: LineSegment[] = [
    { p1: trajectoryPoints[0], p2: trajectoryPoints[1] },
    { p1: trajectoryPoints[1], p2: trajectoryPoints[2] },
  ];

  console.log('Trajectory segments:');
  trajectorySegments.forEach((seg, i) => {
    console.log(
      `  Segment ${i + 1}: (${seg.p1.lon}, ${seg.p1.lat}) → (${seg.p2.lon}, ${seg.p2.lat})`
    );
  });
  console.log('');

  const intersections: Array<{
    cableName: string;
    cableId: string;
    segmentIndex: number;
    trajectorySegment: number;
    intersectionPoint?: Point;
  }> = [];

  const nearbyCables: Array<{
    cableName: string;
    cableId: string;
    minDistance: number;
    closestPoint: Point;
  }> = [];

  // Check each cable
  for (const feature of infrastructureData.features) {
    const cableName = feature.properties.name;
    const cableId = feature.properties.id;
    const geometry = feature.geometry;

    if (geometry.type !== 'MultiLineString') {
      continue;
    }

    // Check each line string in the MultiLineString
    for (let segIdx = 0; segIdx < geometry.coordinates.length; segIdx++) {
      const lineString = geometry.coordinates[segIdx];

      // Check each segment in the line string
      for (let i = 0; i < lineString.length - 1; i++) {
        const cableSegment: LineSegment = {
          p1: { lon: lineString[i][0], lat: lineString[i][1] },
          p2: { lon: lineString[i + 1][0], lat: lineString[i + 1][1] },
        };

        // Check intersection with each trajectory segment
        for (let trajIdx = 0; trajIdx < trajectorySegments.length; trajIdx++) {
          const trajSegment = trajectorySegments[trajIdx];

          if (segmentsIntersect(trajSegment, cableSegment)) {
            intersections.push({
              cableName,
              cableId,
              segmentIndex: segIdx,
              trajectorySegment: trajIdx + 1,
            });
          }
        }

        // Also check if trajectory points are near this cable segment
        for (const trajPoint of trajectoryPoints) {
          if (pointNearSegment(trajPoint, cableSegment, 0.05)) {
            const dist = distance(trajPoint, cableSegment.p1);
            nearbyCables.push({
              cableName,
              cableId,
              minDistance: dist,
              closestPoint: trajPoint,
            });
          }
        }
      }
    }
  }

  // Report results
  console.log('📊 Results:\n');

  if (intersections.length > 0) {
    console.log(`✅ Found ${intersections.length} intersection(s):\n`);
    const uniqueCables = new Map<string, typeof intersections[0]>();
    intersections.forEach((intersection) => {
      const key = `${intersection.cableId}-${intersection.segmentIndex}`;
      if (!uniqueCables.has(key)) {
        uniqueCables.set(key, intersection);
      }
    });

    uniqueCables.forEach((intersection) => {
      console.log(`  🔴 ${intersection.cableName} (${intersection.cableId})`);
      console.log(
        `     Intersects with trajectory segment ${intersection.trajectorySegment}`
      );
      console.log(`     Cable segment index: ${intersection.segmentIndex}\n`);
    });
  } else {
    console.log('❌ No direct intersections found.\n');
  }

  // Check for nearby cables (within 0.05 degrees ≈ 5.5 km)
  const uniqueNearby = new Map<string, typeof nearbyCables[0]>();
  if (nearbyCables.length > 0) {
    console.log(`⚠️  Found ${nearbyCables.length} nearby cable(s) (within ~5.5 km):\n`);
    nearbyCables.forEach((cable) => {
      const key = cable.cableId;
      if (!uniqueNearby.has(key) || uniqueNearby.get(key)!.minDistance > cable.minDistance) {
        uniqueNearby.set(key, cable);
      }
    });

    uniqueNearby.forEach((cable) => {
      console.log(`  ⚠️  ${cable.cableName} (${cable.cableId})`);
      console.log(`     Minimum distance: ~${(cable.minDistance * 111).toFixed(2)} km\n`);
    });
  }

  return {
    intersections: Array.from(uniqueCables.values()),
    nearbyCables: Array.from(uniqueNearby.values()),
  };
}

// Run the check
checkIntersections()
  .then((result) => {
    console.log('\n✅ Analysis complete.');
    process.exit(0);
  })
  .catch((error) => {
    console.error('❌ Error:', error);
    process.exit(1);
  });

